packages/flutter_tools/lib/src/features.dart

```diff
@@ -0,0 +1,245 @@
+/// use this class should extent instead of implement, so that new flags are
+/// picked up automatically.
+class FeatureFlags {
+  const FeatureFlags();
+
+  /// Whether flutter web is enabled.
+  ///
+  /// Defaults to `false`.
+  bool get isWebEnabled => false;
+
+  /// Whether flutter desktop for macOS is enabled.
+  ///
+  /// Defaults to `false`.
+  bool get isMacOSEnabled => false;
+
+  /// Whether flutter desktop for Windows is enabled.
+  ///
+  /// Defaults to `false`.
+  bool get isWindowsEnabled => false;
+
+  /// Whether flutter desktop for linux is enabled.
+  ///
+  /// Defaults to `false`.
+  bool get isLinuxEnabled => false;
+}
+
+/// All current Flutter feature flags.
+const List<Feature> allFeatures = <Feature>[
+  flutterWebFeature,
+  flutterLinuxDesktopFeature,
+  flutterMacOSDesktopFeature,
+  flutterWindowsDesktopFeature,
+];
+
+/// The [Feature] for flutter web.
+const Feature flutterWebFeature = Feature(
+  name: 'Flutter Web',
+  configSetting: 'enable-web',
+  environmentOverride: 'FLUTTER_WEB',
+  master: FeatureSetting(
+    available: true,
+    enabledByDefault: false,
+  ),
+  dev: FeatureSetting(
+    available: true,
+    enabledByDefault: false,
+  ),
+);
+
+/// The [Feature] for macOS desktop.
+const Feature flutterMacOSDesktopFeature = Feature(
+  name: 'Flutter Desktop for macOS',
+  configSetting: 'enable-macos-desktop',
+  environmentOverride: 'ENABLE_FLUTTER_DESKTOP',
+  master: FeatureSetting(
+    available: true,
+    enabledByDefault: false,
+  ),
+);
+
+/// The [Feature] for Linux desktop.
+const Feature flutterLinuxDesktopFeature = Feature(
+  name: 'Flutter Desktop for Linux',
+  configSetting: 'enable-linux-desktop',
+  environmentOverride: 'ENABLE_FLUTTER_DESKTOP',
+  master: FeatureSetting(
+    available: true,
+    enabledByDefault: false,
+  ),
+);
+
+/// The [Feature] for Windows desktop.
+const Feature flutterWindowsDesktopFeature = Feature(
+  name: 'Flutter Desktop for Windows',
+  configSetting: 'enable-windows-desktop',
+  environmentOverride: 'ENABLE_FLUTTER_DESKTOP',
+  master: FeatureSetting(
+    available: true,
+    enabledByDefault: false,
+  ),
+);
+
+/// A [FeatureFlags] that looks up values based on [Feature] definitions.
+class FlutterFeatureFlags implements FeatureFlags {
+  const FlutterFeatureFlags();
+
+  @override
+  bool get isLinuxEnabled => _isEnabled(flutterLinuxDesktopFeature);
+
+  @override
+  bool get isMacOSEnabled => _isEnabled(flutterMacOSDesktopFeature);
+
+  @override
+  bool get isWebEnabled => _isEnabled(flutterWebFeature);
+
+  @override
+  bool get isWindowsEnabled => _isEnabled(flutterWindowsDesktopFeature);
+
+  // Calculate whether a particular feature is enabled for the current channel.
+  static bool _isEnabled(Feature feature) {
+    final String currentChannel = FlutterVersion.instance.channel;
+    final FeatureSetting featureSetting = feature.getSettingForChannel(currentChannel);
+    if (!featureSetting.available) {
+      return false;
```

Comment:
How do we surface this to the user if the user tries to turn something on not available on the channel?

Reply:
I've added some help text to the config setting if the feature is currently unavailable. The command line help option also includes which channels it can be used with.